File: YYVolumeManager/YYVolumeManager/Source/YYVolumeManager.m
//
//  YYVolumeManager.m
//  Lix
//
//  Created by Phoenix on 2017/9/5.
//  Copyright © 2017年 Phoenix. All rights reserved.
//

#import "YYVolumeManager.h"
#import <MediaPlayer/MediaPlayer.h>
#import <AVFoundation/AVFoundation.h>

@interface YYVolumeManager()
@property (nonatomic, strong) MPVolumeView *defaultVolumeView;
@property (nonatomic, strong) UISlider *volumeSlider;
@end

@implementation YYVolumeManager {
    NSHashTable *_observers;
}

static YYVolumeManager *_instance;
static dispatch_once_t onceToken;

+ (YYVolumeManager *)shared {
    dispatch_once(&onceToken, ^{
        _instance = [[super alloc] _init];
    });
    return _instance;
}

+ (void)releaseSingleton {
    [_instance removeNotification];
    onceToken = 0;
    _instance = nil;
}

- (void)dealloc {
#if DEBUG
    NSLog(@"YYVolumeManager dealloc");
#endif
}

- (instancetype)_init {
    self = [super init];
    _observers = [[NSHashTable alloc] initWithOptions:NSPointerFunctionsWeakMemory | NSPointerFunctionsObjectPersonality capacity:0];
    _volume = [[AVAudioSession sharedInstance] outputVolume];
    if ([NSThread isMainThread]) {
        self.defaultVolumeView = [[MPVolumeView alloc] init];
    } else {
        dispatch_sync(dispatch_get_main_queue(), ^{
            self.defaultVolumeView = [[MPVolumeView alloc] init];
        });
    }
    [self registeNotification];
    return self;
}

- (void)addObserver:(id<YYVolumeObserver>)observer {
    if (!observer) {
        return;
    }
    [_observers addObject:observer];
}

- (void)removeObserver:(id<YYVolumeObserver>)observer {
    if (!observer) {
        return;
    }
    [_observers removeObject:observer];
}

- (void)setCustomVolumeUI:(BOOL)customVolumeUI {
    [self.defaultVolumeView removeFromSuperview];
    if (customVolumeUI) {
        [self.defaultVolumeView setFrame:CGRectMake(-100, -100, 40, 40)];
        [[UIApplication sharedApplication].keyWindow addSubview:self.defaultVolumeView];
    }
}


- (UISlider *)volumeSlider {
    if (!_volumeSlider) {
        for (UIView *view in self.defaultVolumeView.subviews) {
            if ([view.class.description isEqualToString:@"MPVolumeSlider"]){
                _volumeSlider = (UISlider *)view;
                break;
            }
        }
    }
    return _volumeSlider;
}

- (void)setVolume:(CGFloat)volume {
    _volume = volume;
    [self.volumeSlider setValue:volume
                       animated:NO];
}

- (void)registeNotification {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(volumeChanged:)
                                                 name:@"AVSystemController_SystemVolumeDidChangeNotification"
                                               object:nil];
    
    [[UIApplication sharedApplication] beginReceivingRemoteControlEvents];
}

- (void)volumeChanged:(NSNotification *)notification {
    CGFloat currentVolume = [[[notification userInfo] objectForKey:@"AVSystemController_AudioVolumeNotificationParameter"] floatValue];
    self.volume = currentVolume;

    for (id<YYVolumeObserver> observer in _observers.copy) {
        if (observer && [observer respondsToSelector:@selector(volumeChanged:)]) {
            [observer volumeChanged:currentVolume];
        }
    }
}

- (void)removeNotification {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [[UIApplication sharedApplication] endReceivingRemoteControlEvents];
}
@end

File: Swift/YYVolumeManager/Source/YYVolumeManager.swift
//
//  YYVolumeManager.swift
//  YYVolumeManager
//
//  Created by Phoenix on 2019/2/19.
//  Copyright © 2019 Phoenix. All rights reserved.
//

import Foundation
import MediaPlayer
import AVFoundation
import UIKit

@objc public protocol YYVolumeObserver {
    @objc optional func volumeChanged(value: Float)
}

public final class YYVolumeManager {
    public static let shared = YYVolumeManager()
    
    public var volume: Float {
        didSet {
            self.volumeSlider.setValue(volume, animated: false)
        }
    }
    
    public var customVolumeUI: Bool {
        didSet {
            self.defaultVolumeView.removeFromSuperview()
            if customVolumeUI {
                self.defaultVolumeView.frame = CGRect(x: -100, y: -100, width: 40, height: 40)
                UIApplication.shared.keyWindow!.addSubview(self.defaultVolumeView)
            }
        }
    }
    
    private var observers: NSHashTable<AnyObject>
    private var defaultVolumeView: MPVolumeView
    private lazy var volumeSlider: UISlider = {
        for view in self.defaultVolumeView.subviews {
            if String(describing: type(of: view)) == "MPVolumeSlider" {
                return view as! UISlider
            }
        }
        return UISlider()
    }()
    
    private convenience init() {
        self.init(volume: AVAudioSession.sharedInstance().outputVolume)
    }
    
    private init(volume: Float, customVolumeUI: Bool = false) {
        self.volume = volume
        self.customVolumeUI = customVolumeUI
        observers = NSHashTable(options: [.weakMemory, .objectPersonality], capacity: 0)
        defaultVolumeView = MPVolumeView()
        registeNotification()
    }
    
    public func addObserver(_ observer: YYVolumeObserver) {
        self.observers.add(observer as AnyObject)
    }
    
    public func removeObserver(_ observer: YYVolumeObserver) {
        self.observers.remove(observer as AnyObject)
    }
    
    private func registeNotification() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(volumeChanged),
            name: NSNotification.Name(rawValue: "AVSystemController_SystemVolumeDidChangeNotification"),
            object: nil
        )
        UIApplication.shared.beginReceivingRemoteControlEvents()
    }
    
    private func removeNotification() {
        NotificationCenter.default.removeObserver(self)
        UIApplication.shared.endReceivingRemoteControlEvents()
    }
    
    @objc private func volumeChanged(notification: Notification) {
        guard let userInfo = notification.userInfo else {
            return
        }
        if let currentVolume = userInfo["AVSystemController_AudioVolumeNotificationParameter"] as? Float {
            self.volume = currentVolume
            let enumerator = self.observers.objectEnumerator()
            
            while let observer = enumerator.nextObject() as AnyObject? {
                if observer.responds(to: #selector(YYVolumeObserver.volumeChanged)) {
                    observer.volumeChanged(value: currentVolume)
                }
            }
        }
    }
    
}

